File: 2021/src/main/scala/Day16.scala
import scala.io.Source
import cats._
import cats.implicits._
import cats.data.State
import scala.annotation.tailrec
import scala.util.chaining._

@main def Day16 = {

  /** **********************************************
    * Import data
    * **********************************************
    */
  val input = Source.fromFile("2021/src/main/resources/day16.txt").getLines.toList.head
  val binaries = input.toList
    .flatMap(c => "%04d".format(c.toString.pipe(Integer.parseInt(_, 16)).toBinaryString.toInt))
    .mkString

  /** **********************************************
    * Prepare
    * **********************************************
    */
  trait Content
  case class Operator(version: Long, typeId: Long, data: List[Content]) extends Content
  case class Operand(version: Long, typeId: Long, data: Long) extends Content

  object Parser {

    type Parser[A] = State[String, A]
    def toLong(b: String) = BigInt(b, 2).toLong
    def peekSize = State[String, Int] { s => (s, s.size) }
    def take(n: Int) = State[String, String] { s => if (s.size < n) "" -> s else s.splitAt(n).swap }

    // Parse single package
    def parse: Parser[Content] =
      for {
        // Retrieve header info
        version <- take(3).map(toLong)
        typeId  <- take(3).map(toLong)

        // Define length of subpackages if applicable
        id <-
          if (typeId == 4) take(0).map(_ => -1L)
          else take(1).map(toLong)
        n <-
          if (id == 0) take(15).map(toLong).map(_.toInt)
          else if (id == 1) take(11).map(toLong).map(_.toInt)
          else take(0).map(_ => -1)
        // Get content
        len <- peekSize
        pkg <-
          if (id == 0) parseSize(len - n).map(d => Operator(version, typeId, d))
          else if (id == 1) parseN(n).map(d => Operator(version, typeId, d))
          else parseData.map(b => Operand(version, typeId, toLong(b)))
      } yield pkg

    // Parse literal in binary format
    def parseData: Parser[String] = for {
      head <- take(5)
      tail <- if (head.head == '1') parseData else take(0).map(_ => "")
    } yield (head.tail + tail)

    // Parse specific count of subpackages
    def parseN(n: Int): Parser[List[Content]] = for {
      head <- parse
      tail <- if (n > 1) parseN(n - 1) else take(0).map(_ => Nil)
    } yield (head :: tail)

    // Parse specific length of bits
    def parseSize(n: Int): Parser[List[Content]] = for {
      head <- parse
      len  <- peekSize
      tail <-
        if (len > n) parseSize(n)
        else take(0).map(_ => Nil)
    } yield (head :: tail)
  }

  /** **********************************************
    * Process
    * **********************************************
    */
  def process(pkg: Content): Long = pkg match
    case Operand(_, _, value) => value.toLong
    case Operator(_, typeId, data) =>
      val values = data.map(process)
      typeId match
        case 0 => values.reduce(_ + _)
        case 1 => values.reduce(_ * _)
        case 2 => values.min
        case 3 => values.max
        case 4 => values.sum // will be taken care of in case Data.
        case 5 => values.pipe { l => if (l(0) > l(1)) 1 else 0 }
        case 6 => values.pipe { l => if (l(0) < l(1)) 1 else 0 }
        case 7 => values.pipe { l => if (l(0) == l(1)) 1 else 0 }

  val content = Parser.parse.runA(binaries).value
  val result = process(content)

  /** **********************************************
    * Output
    * **********************************************
    */
  def show(result: Long) =
    println(f"Result: $result")

  show(result)
}

File: 2021/src/main/python/day16.py
from dataclasses import dataclass
from functools import reduce

#################################################
# Import
#################################################
with open("2021/src/main/resources/day16.txt", "r") as f:
    data = f.readline().strip()
    data = "".join(["{:0>4b}".format(int(h, 16)) for h in list(data)])

#################################################
# Prepare
#################################################
@dataclass
class Package():
    version: int
    type_id: int
    id: int
    value: int
    children: list

@dataclass
class Stream():
    data: str
    def __init__(self, data: str):
        self.data = data

    def take(self, i: int):
        msg = self.data[0:i]
        self.data = self.data[i:]
        return msg
    
    def peek(self, i: int):
        if len(self.data) < i:
            return self.data
        else:
            return self.data[0:i]

    def __len__(self):
        return len(self.data)


def parse(stream: Stream):
    version = int(stream.take(3), 2)
    type_id = int(stream.take(3), 2)
    id = stream.peek(1)
    if type_id == 4:
        # parse literal
        value = ""
        bit = "1"
        while bit[0] == "1":
            bit = stream.take(5)
            value += bit[1:]
        return Package(version, type_id, None, int(value, 2), [])
    elif id == "0":
        # parse by size
        _ = stream.take(1)
        size = int(stream.take(15), 2)
        substream = Stream(stream.take(size))
        content = []
        while len(substream) > 0:
            content.append(parse(substream))
        return Package(version, type_id, id, None, content)
    else:
        # Parse several times
        _ = stream.take(1)
        n = int(stream.take(11), 2)
        content = []
        for _ in range(n):
            content.append(parse(stream))
        return Package(version, type_id, id, None, content)


def calc_value(pkg: Package) -> int:
    fun = {
        0: lambda x, y: x + y,
        1: lambda x, y: x * y,
        2: min,
        3: max,
        5: lambda x, y: 1 if x > y else 0,
        6: lambda x, y: 1 if x < y else 0,
        7: lambda x, y: 1 if x == y else 0
    }
    if pkg.type_id == 4:
        return pkg.value
    else:
        return reduce(fun[pkg.type_id], [calc_value(kid) for kid in pkg.children])

#################################################
# Process
#################################################
content = parse(Stream(data))

#################################################
# Output
#################################################
print(f"{calc_value(content) = }")
